ranking/src/ts_system/mod.rs

```diff
@@ -0,0 +1,314 @@
+mod normal;
+mod nodes;
+
+use super::contest_config::Contest;
+use super::compute_ratings::{RatingSystem, Player, Rating};
+
+use nodes::{ProdNode, LeqNode, GreaterNode, SumNode, TreeNode, ValueNode, FuncNode};
+use normal::Gaussian;
+
+use rayon::prelude::*;
+use std::cell::{RefCell, RefMut};
+use std::cmp::max;
+use std::collections::{HashMap, VecDeque};
+use std::collections::hash_map::Entry;
+
+use std::rc::{Rc, Weak};
+use std::f64::INFINITY;
+
+type TSPlayerRating = Gaussian;
+type TSMessage = nodes::Message;
+type TSPlayer = String;
+type TSTeam = Vec<TSPlayer>;
+type TSContestPlace = Vec<TSTeam>;
+type TSContest = Vec<TSContestPlace>;
+type TSRating = HashMap<TSPlayer, TSPlayerRating>;
+
+// TrueSkillStPB rating system
+pub struct TrueSkillSPBSystem {
+    // epsilon used for ties
+    eps: f64,
+    // default player rating
+    mu: f64,
+    // default player sigma
+    sigma: f64,
+    // performance sigma
+    beta: f64,
+    // epsilon used for convergence loop
+    convergence_eps: f64,
+    // defines sigma growth per second
+    sigma_growth: f64,
+    // history of ratings for all players
+    rating: TSRating,
+}
+
+impl Default for TrueSkillSPBSystem {
+    fn default() -> Self {
+        Self {
+            eps: 0.70,
+            mu: 1500.,
+            sigma: 1500. / 3., // mu/3
+            beta: 1500. / 6., // sigma/2
+            convergence_eps: 2e-4,
+            sigma_growth: 5.,
+            rating: TSRating::new(),
+        }
+    }
+}
+
+// fn update_rating(old: &TSRating, new: &mut TSRatingHistory, contest: &TSContest, when: usize) {
+//     for place in &contest[..] {
+//         for team in &place[..] {
+//             for player in &team[..] {
+//                 new.entry(player.clone()).or_insert(Vec::new()).push((old.get(player).unwrap().clone(), when));
+//             }
+//         }
+//     }
+// }
+
+
+fn gen_team_message<T, K: Clone>(places: &Vec<Vec<T>>, default: &K) -> Vec<Vec<K>> {
+    let mut ret: Vec<Vec<K>> = Vec::with_capacity(places.len());
+
+    for place in places {
+        ret.push(vec![default.clone(); place.len()]);
+    }
+
+    ret
+}
+
+
+fn gen_player_message<T, K: Clone>(places: &Vec<Vec<Vec<T>>>, default: &K) -> Vec<Vec<Vec<K>>> {
+    let mut ret = Vec::with_capacity(places.len());

```

Comment:
`places.iter().map(|place| place.iter().map(|team| vec![default.clone(); team.len()]).collect()).collect()`

(`cargo fmt` will help with the indents and formatting)